ocaml/xapi/xapi_vm_group_helpers.ml

```diff
@@ -0,0 +1,208 @@
+(*
+ * Copyright (C) 2024 Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = "xapi_vm_group_helpers" end)
+
+open D
+
+let check_breach_on_vm_anti_affinity_rules ~__context ~group =
+  Db.VM_group.get_VMs ~__context ~self:group
+  |> List.filter_map (fun vm ->
+         let vm_rec = Db.VM.get_record ~__context ~self:vm in
+         match (vm_rec.API.vM_power_state, vm_rec.API.vM_resident_on) with
+         | `Running, h when h <> Ref.null ->
+             Some h
+         | _ ->
+             None
+     )
+  |> function
+  | [] | [_] ->
+      None
+  | h :: remaining ->
+      if List.exists (fun h' -> h' <> h) remaining then
+        None
+      else
+        Some h
+
+let report_anti_affinity_alert ~__context ~group ~host =
+  let group_uuid = Db.VM_group.get_uuid ~__context ~self:group in
+  let host_uuid = Db.Host.get_uuid ~__context ~self:host in
+  let body =
+    Printf.sprintf
+      "<body><message>Breach on VM anti-affinity rules</message><VM_group>"
+    ^ group_uuid
+    ^ "</VM_group><host>"
+    ^ host_uuid
+    ^ "</host></body>"
+  in
+  let obj_uuid =
+    Db.Pool.get_uuid ~__context ~self:(Helpers.get_pool ~__context)
+  in
+  Xapi_alert.add
+    ~msg:
+      Api_messages
+      .all_vms_in_the_anti_affinity_group_are_running_on_a_single_host
+    ~cls:`Pool ~obj_uuid ~body
+
+let get_anti_affinity_alerts ~__context =
+  Helpers.call_api_functions ~__context (fun rpc session_id ->
+      Client.Client.Message.get_all_records ~rpc ~session_id
```

Comment:
`Db.Message.get_all_records` should work as well.

Reply:
I encounter error: `Error: Unbound module Db.Message` , 
the message is saved as local file, I am not sure if we can get it from DB?

Reply:
Indeed. `~in_db:false`.
Sorry for this wrong comment.